CMakeLists.txt

```diff
@@ -1,149 +1,351 @@
+
+set(KISSFFT_DATATYPE "float" CACHE STRING "Principal datatype of kissfft: double, float (default), int16_t, int32_t, simd")
+
+# Additional options
+
+option(KISSFFT_OPENMP "Build kissfft with OpenMP support" OFF)
+option(KISSFFT_PKGCONFIG "Build pkg-config files" ON)
+option(KISSFFT_STATIC "Build kissfft as static (ON) or shared library (OFF)" OFF)
+option(KISSFFT_TEST "Build kissfft tests" ON)
+option(KISSFFT_TOOLS "Build kissfft command-line tools" ON)
+option(KISSFFT_USE_ALLOCA "Use alloca instead of malloc" OFF)
+
+#
+# Validate datatype
+#
+
+if (NOT KISSFFT_DATATYPE MATCHES "^double$" AND
+    NOT KISSFFT_DATATYPE MATCHES "^float$" AND
+    NOT KISSFFT_DATATYPE MATCHES "^int16_t$" AND
+    NOT KISSFFT_DATATYPE MATCHES "^int32_t$" AND
+    NOT KISSFFT_DATATYPE MATCHES "^simd$")
+    message(FATAL_ERROR "Incorrect value of KISSFFT_DATATYPE! It can be one of: double, float, int16_t, int32_t, simd")
+endif()
+
+#
+# Print principal datatype
+#
+
+message(STATUS "Building KissFFT with datatype: ${KISSFFT_DATATYPE}")
+set(KISSFFT_OUTPUT_NAME "kissfft-${KISSFFT_DATATYPE}")
+
+#
+# Validate KISSFFT_STATIC
+#
+
+if (BUILD_SHARED_LIBS AND KISSFFT_STATIC)
+    message(FATAL_ERROR "Conflicting CMake configuration: -DBUILD_SHARED_LIBS=ON and -DKISSFFT_STATIC=ON")
 endif()
 
+#
+# Enable BUILD_SHARED_LIBS for shared library build before
+# kissfft library is declared
+#
+
+if (NOT KISSFFT_STATIC)
+    set(BUILD_SHARED_LIBS ON)
+    message(STATUS "Building shared library")
+else()
+    message(STATUS "Building static library")
+endif()
+
+#
+# Detect C compiler and pass appropriate flags
+#
+
 if(CMAKE_C_COMPILER_ID MATCHES "GNU|Clang|AppleClang")
     add_compile_options(-ffast-math -fomit-frame-pointer
         -W -Wall -Wcast-align -Wcast-qual -Wshadow -Wwrite-strings
         "$<$<COMPILE_LANGUAGE:C>:-Wstrict-prototypes;-Wmissing-prototypes;-Wnested-externs;-Wbad-function-cast>")
 endif()
 
-set(KISSFFT_FIXED_WIDTH_VALID 16 32)
-set(KISSFFT_FIXED_WIDTH "16" CACHE STRING "Width of kissfft fixed integer (16/32)")
-set_property(CACHE KISSFFT_FIXED_WIDTH PROPERTY STRINGS ${KISSFFT_FIXED_WIDTH_VALID})
-if(NOT KISSFFT_FIXED_WIDTH IN_LIST KISSFFT_FIXED_WIDTH_VALID)
-    message(FATAL_ERROR "KISSFFT_FIXED_WIDTH must be 16 or 32")
+#
+# Add GNUInstallDirs for GNU infrastructure before target)include_directories
+#
+
+if(CMAKE_SYSTEM_NAME MATCHES "^(Linux|kFreeBSD|GNU)$" AND NOT CMAKE_CROSSCOMPILING)
+    include(GNUInstallDirs)
 endif()
 
-set(KISSFFT_FLOAT_TYPE_VALID float double)
-set(KISSFFT_FLOAT_TYPE "float" CACHE STRING "Floating type of kissfft (float/double)")
-set_property(CACHE KISSFFT_FLOAT_TYPE PROPERTY STRINGS ${KISSFFT_FLOAT_TYPE_VALID})
-if(NOT KISSFFT_FLOAT_TYPE IN_LIST KISSFFT_FLOAT_TYPE_VALID)
-    message(FATAL_ERROR "KISSFFT_FLOAT_TYPE must be float or double")
+#
+# Declare CMAKE_INSTALL_{FULL_,}PKGINCLUDEDIR for kissfft include path
+#
+
+if (NOT DEFINED CMAKE_INSTALL_PKGINCLUDEDIR)
```

Comment:
This can be made unconditional. And probably should avoid the `CMAKE_INSTALL_` prefix since it does not come from CMake.

Reply:
I think `INSTALL_PKGINCLUDEDIR` is fair name reflecting that it does not come from CMake internals but also not coming from `PKGINCLUDEDIR` environment variable if specified. What do you think?

Reply:
Sounds okay to me too. Though I have not seen `PKGINCLUDEDIR` environment variable being used and I would not expect it to be a problem. Or do CMake or GNU Make just pick up random env vars?

Reply:
> Or do CMake or GNU Make just pick up random env vars?

In CMake, one explicitly needs to ask `$ENV{NAME}`, GNU make picks all them and passes the `export`-ed ones to submakes.